enhancements/builds/volume-mounted-resources.md

```diff
@@ -0,0 +1,297 @@
+  // the docker image should be pulled
+  From kapi.ObjectReference
+
+  // PullSecret is the name of a Secret that would be used for setting up
+  // the authentication for pulling the Docker images from the private Docker
+  // registries
+  PullSecret *kapi.LocalObjectReference
+
+  // Env contains additional environment variables you want to pass into a builder container.
+  Env []kapi.EnvVar
+
+  // Scripts is the location of Source scripts
+  Scripts string
+
+  // Incremental flag forces the Source build to do incremental builds if true.
+  Incremental *bool
+
+  // ForcePull describes if the builder should pull the images from registry prior to building.
+  ForcePull bool
+
+  // Volumes is a list of volumes that can be mounted by the build
+  // Only a subset of Kubernetes Volume sources are supported by builds.
+  // More info: https://kubernetes.io/docs/concepts/storage/volumes
+  Volumes []BuildVolume
+
+  // VolumeMounts are volumes to mount into the build
+  VolumeMounts []VolumeMount
+}
+```
+
+```go
+// BuildVolume describes a volume that is made available to build pods, such that it can be mounted into the build execution environment.
+// Only a subset of Kubernetes Volume sources are supported by builds.
+type BuildVolume struct {
+
+  // Volume's name.
+	// Must be a DNS_LABEL and unique within the pod.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// BuildVolumeSource represents the location and type of the mounted volume.
+	BuildVolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volumeSource"`
+}
+
+
+// Represents the source of a volume to mount.
+// Only one of its members may be specified.
+// This is a subset of the core Kubernetes VolumeSource.
+type BuildVolumeSource struct {
+	// Secret represents a secret that should populate this volume.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#secret
+	// +optional
+	Secret *kapi.SecretVolumeSource `json:"secret,omitempty" protobuf:"bytes,1,opt,name=secret"`
+	
+	// ConfigMap represents a configMap that should populate this volume
+	// +optional
+	ConfigMap *kapi.ConfigMapVolumeSource `json:"configMap,omitempty" protobuf:"bytes,2,opt,name=configMap"`
+
+	// CSI (Container Storage Interface) represents ephemeral storage that is handled by certain external CSI drivers (Beta feature).
+	// +optional
+	CSI *kapi.CSIVolumeSource `json:"csi,omitempty" protobuf:"bytes,3,opt,name=csi"`
+}
+```
+
+```go
+// VolumeMount describes a mounting of a Volume within the build execution environment.
+type VolumeMount struct {
+  // This must match the Name of a Volume.
+  Name string
+  // Mounted read-only if true, read-write otherwise (false or unspecified).
+  // Defaults to false.
+  // +optional
+  ReadOnly bool
+  // Path within the build environment at which the volume should be mounted.  Must
+  // not contain ':'.
+  MountPath string
```

Comment:
So I almost posed this question above, but chose to do it here with the literal struct golang definition.

I'll copy the text from above for reference:

```
These fields will be wired, via the build controller, directly to the build pod that is constructed.
In addition all mounts into the pod will be done at a path of our choosing, not the VolumeMount path specified, to ensure the user cannot overwrite critical function inside the build pod and use it as an escalation pathway.
Example - all volumes are mounted under `/var/run/openshift.io/volumes` in the build pod containers.
```

So I'm trying to reconcile those words above, including `mounts into the pod will be done at a path of our choosing, not the VolumeMount path specified` with this presence of this field.

Is this a relative path off of `/var/run/openshift.io/volumes` then?  Or are we talking pod file system vs. build container context blah blah blah sort of thing, and this means transient mounts when we call buildah from openshift/builder ?

If so, use of `relative path` and `build context`  in the godoc above seems needed to me to make this clear.

If it is something other then relative path, I need more explanation in general.

Reply:
>  Or are we talking pod file system vs. build container context blah blah blah sort of thing, and this means transient mounts when we call buildah from openshift/builder ?

yeah, it's making that distinction.  We will choose the mount location within the build pod.  The user will choose the mount location within the buildah spawned container (the one that's executing the Dockerfile commands).

Reply:
Proposal and API doc updated to distinguish "pod volume mount" from "buildah volume mount"